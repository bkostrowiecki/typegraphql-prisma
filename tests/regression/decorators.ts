import { promises as fs } from "fs";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";
import createReadGeneratedFile, {
  ReadGeneratedFile,
} from "../helpers/read-file";

describe("decorators", () => {
  let outputDirPath: string;
  let readGeneratedFile: ReadGeneratedFile;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath("regression-decorators");
    await fs.mkdir(outputDirPath, { recursive: true });
    readGeneratedFile = createReadGeneratedFile(outputDirPath);
  });

  describe("on model", () => {
    it("should generate decorator on model", async () => {
      const schema = /* prisma */ `
      /// @@TypeGraphQL.decorator(name: "TypeGraphQL.Authorized", args: "'admin'")
      model User {
        intIdField  Int     @id @default(autoincrement())
      }
    `;

      await generateCodeFromSchema(schema, { outputDirPath });
      const userModelTSFile = await readGeneratedFile("/models/User.ts");

      expect(userModelTSFile).toMatchSnapshot("User");
    });

    it("should generate decorator on model with one argument admin", async () => {
      const schema = /* prisma */ `
      /// @@TypeGraphQL.decorator(name: "TypeGraphQL.Authorized", args: "['admin']")
      model User {
        intIdField  Int     @id @default(autoincrement())
      }
    `;

      await generateCodeFromSchema(schema, { outputDirPath });
      const userModelTSFile = await readGeneratedFile("/models/User.ts");

      expect(userModelTSFile).toMatchSnapshot("User");
    });

    it("should generate decorator on model with two arguments admin and participant", async () => {
      const schema = /* prisma */ `
      /// @@TypeGraphQL.decorator(name: "TypeGraphQL.Authorized", args: ["'admin'", "'participant'"])
      model User {
        intIdField  Int     @id @default(autoincrement())
      }
    `;

      await generateCodeFromSchema(schema, { outputDirPath });
      const userModelTSFile = await readGeneratedFile("/models/User.ts");

      expect(userModelTSFile).toMatchSnapshot("User");
    });
  });

  describe("on field", () => {
    it("should generate decorator on field", async () => {
      const schema = /* prisma */ `
      model User {
        /// @@TypeGraphQL.decorator(name: "TypeGraphQL.Authorized", args: "'admin'")
        intIdField  Int     @id @default(autoincrement())
      }
    `;

      await generateCodeFromSchema(schema, { outputDirPath });
      const userModelTSFile = await readGeneratedFile("/models/User.ts");

      expect(userModelTSFile).toMatchSnapshot("User");
    });
  });
});
