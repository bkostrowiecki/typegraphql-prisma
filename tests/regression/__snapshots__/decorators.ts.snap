// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators on field should generate decorator on field: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  @TypeGraphQL.Authorized('admin')
  intIdField!: number;
}
"
`;

exports[`decorators on model should generate decorator on model with one argument admin: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
@TypeGraphQL.Authorized(['admin'])
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  intIdField!: number;
}
"
`;

exports[`decorators on model should generate decorator on model with two arguments admin and participant: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
@TypeGraphQL.Authorized('admin', 'participant')
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  intIdField!: number;
}
"
`;

exports[`decorators on model should generate decorator on model: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
@TypeGraphQL.Authorized('admin')
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  intIdField!: number;
}
"
`;

exports[`decorators should generate decorator on model with one argument admin: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
@TypeGraphQL.Authorized(['admin'])
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  intIdField!: number;
}
"
`;

exports[`decorators should generate decorator on model with one argument as an array: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
@TypeGraphQL.Authorized(admin, participant)
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  intIdField!: number;
}
"
`;

exports[`decorators should generate decorator on model with two arguments admin and participant: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
@TypeGraphQL.Authorized('admin', 'participant')
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  intIdField!: number;
}
"
`;

exports[`decorators should generate decorator on model: User 1`] = `
"import * as TypeGraphQL from \\"type-graphql\\";
import * as GraphQLScalars from \\"graphql-scalars\\";
import { Prisma } from \\"../../../helpers/prisma-client-mock\\";
import { DecimalJSScalar } from \\"../scalars\\";

@TypeGraphQL.ObjectType(\\"User\\", {})
@TypeGraphQL.Authorized('admin')
export class User {
  @TypeGraphQL.Field(_type => TypeGraphQL.Int, {
    nullable: false
  })
  intIdField!: number;
}
"
`;
